### ISSUE TITLE
RFC: Compiler support for closures.

### ISSUE THREAD

**Issue by ryzhyk** [vmware/differential-datalog#669]
Closure types:

```
// Variable whose type is closure.
var closure1: |u32, Option<string>|:  u32

// parens can be used for better readability.
var closure2: (|u32, Option<string>|:  u32)

// Alias to a closure type.
typedef Closure<'X> = |'X|: bool
typedef StringClosure = Closure<string>

/*  Some higher-order functions. */
fn map(v: Vec<'X>, f: |'X|: 'Y): Vec<'Y>
fn any(v: Vec<'X>, f: |'X|: bool): bool
fn find(v: Vec<'X>, f: |'X|: bool): Option<'X>

/* A closure can modify its mutable arguments */

// Apply `f` to all elements of a JSON document recursively.
fn map(doc: JsonValue, f: |mut JsonValue|: ())
```

Creating closures:

```
// Closure type inferred from context
|x, y| {
    x + if (y.is_none()) {0} else {5}
}

// The user may type-annotate all or some of the arguments.
|x, y: Option<string>| {
    x + if (y.is_none()) {0} else {5}
}

// Or the returned type.
|x, y: Option<string>|: u32 {
    x + if (y.is_none()) {0} else {5}
}

{
    var json_doc = .....;
    // New attribute to add to each JSON map in the document.
    var new_attr = "bar";
    // When type annotations are provided, the `mut` attribute must be specified too.
    json_doc.map(|v: mut JsonValue| match (v) {
        // The closure can access variables visible in the local scope in read-only mode.
        // Variables are captured using copy semantics, i.e., all local variables accessed by
        // the closure are cloned and stored with the closure.
        JsonMap{m} -> m.insert("foo", JsonString{new_attr}),
        _ -> ()
    })
}
```

@mbudiu-vmw , @blp

**Comment by mbudiu-vmw** [comment]
I think you should call them lambdas, not closures.
Also, I really don't like the Rust syntax. I like the C# and TypeScript syntax much better
`(a: type, b: type) => { body }`
or
`(a: type, b: type) => expression`
The type of a lambda is written as `(a: type) => resultType`. Here we have to be careful; in typescript I think they made the wrong choice, since in `a => resultType` the identifier `a` is not a type, but an argument name! I think it should be the opposite.
In typescript you can also write function types as `function(a: type): resultType`.
See https://www.typescriptlang.org/docs/handbook/functions.html

**Comment by ryzhyk** [comment]
> I think you should call them lambdas, not closures.

I really think they are closures, as they can capture variables from the environment.

> (a: type, b: type) => { body }
> or
> (a: type, b: type) => expression

If we go down this path, the I vote for the 

```
function(t1, .., tn): t
```

syntax, which is consistent with the rest of DDlog syntax and is less ambiguous to parse.  But then how do you write an instance of a closure? Like this?

```
var closure = function(...) {...}
```

**Comment by ryzhyk** [comment]
The downside is that `function` is a long keyword.

**Comment by mbudiu-vmw** [comment]
the latter is the javascript syntax.

**Comment by mbudiu-vmw** [comment]
you can have one syntax to write lambda types, and another one for lambda expressions

**Comment by mbudiu-vmw** [comment]
typescript allows both, for backwards compatibility with js

**Comment by ryzhyk** [comment]
>  (a: type, b: type) => expression

Where does the return type go?

**Comment by mbudiu-vmw** [comment]
This is the expression; this expression has a type of `(a: type, b: type) => expressionType`.

**Comment by ryzhyk** [comment]
> you can have one syntax to write lambda types, and another one for lambda expressions

I think it's good to be consistent.

**Comment by ryzhyk** [comment]
> This is the expression; this expression has a type of (a: type, b: type) => expressionType.

How do I write a closure expression and explicitly specify its return type?

**Comment by ryzhyk** [comment]
I see two options that work so far:

Option 1:
```
var closure: |u32, u32|:  u32;
var closure = |x,y| {x + y}
```

Option 2:
```
var closure: function(u32, u32):  u32;
var closure = function(x: u32, y: u32): u32 {x + y}
```

I like Option 2 because it looks like a natural extension of existing DDlog syntax.
Option 1 can be seen as a short form of option 2 (using `||` instead of `function`), so we can even support both.

**Comment by ryzhyk** [comment]
Implemented in #772 .
